Implement DrawableTextures

ColinSORourke opened the pull request:
https://github.com/godotengine/godot-proposals/issues/7379

[EDITED 9/25 to match completed state of PR]

This PR is an implementation of @reduz's proposal for Drawable Textures in the Godot Engine. Authored primarily by me, with some much appreciated assistance from @clayjohn  and @BastiaanOlij.

See Demo Vid:

https://github.com/user-attachments/assets/95b1f1dc-fa03-4788-b03c-ffad469ede48

Download Demo Project: [drawable-texture-tests.zip](https://github.com/user-attachments/files/22552003/drawable-texture-tests.zip)

The implementation follows Reduz’s specification pretty strictly, with a few departures based on feedback from Rendering Contributor Meetings. Most importantly, we decided to leave implementation of Blit_Polygon functions for a future pull request due to complexity.  We also added a p_material parameter to Blit_Rect & Blit_Rect_Multi on the user exposed resource to allow users to pass their own materials into the resource (if left blank, still defaults to 'default material').

This PR implements:
The Rendering Server API for DrawableTextures 
```cpp
enum TextureDrawableFormat {
   TEXTURE_DRAWABLE_FORMAT_RGBA8,
   TEXTURE_DRAWABLE_FORMAT_RGBA8_SRGB, // Use this if you want to read the result from both 2D (non-hdr) and 3D.
   TEXTURE_DRAWABLE_FORMAT_RGBAH,
   TEXTURE_DRAWABLE_FORMAT_RGBAF,
};

// Create
RID texture_drawable_create(int p_width, int p_height, TextureDrawableFormat p_format, const Color &p_color = Color(1, 1, 1, 1), bool p_with_mipmaps = false);

// Blit a rect
void texture_drawable_blit_rect(const TypedArray<RID> &p_textures, const Rect2i &p_rect, RID p_material, const Color &p_modulate, const TypedArray<RID> &p_source_textures, int p_to_mipmap = 0);

// Utility functions
RID texture_drawable_get_default_material(); // To use with simplified functions in DrawableTexture2D
void texture_drawable_generate_mipmaps(RID p_texture) // Populate the mipmaps for this drawable texture
``` 
The new ShaderType, TextureBlit Shaders
```glsl
shader_type texture_blit
// Optional render modes:
render_mode blend_mix /* default , also available: blend_add, blend_sub, blend_mul */;

uniform sampler2D source_texture : hint_blit_source0;
uniform sampler2D source_texture1 : hint_blit_source1;
// up to 4 sources

void blit() {
     // Read sources and blit
     COLOR0 = texture(source_texture,UV) * MODULATE;
     COLOR1 = texture(source_texture1,UV);
     // up to COLOR3
}
```
And several user-exposed Resource classes to interact with DrawableTextures, most importantly of course, DrawableTexture2D
```cpp
class DrawableTexture2D : public Texture2D {
   enum DrawableFormat {
       DRAWABLE_FORMAT_RGBA8,
       DRAWABLE_FORMAT_RGBA8_SRGB,
       DRAWABLE_FORMAT_RGBAH,
       DRAWABLE_FORMAT_RGBAF,
   };
    void setup(int p_width, int p_height, DrawableFormat p_format, const Color &p_modulate = Color(1, 1, 1, 1), bool p_use_mipmaps = false);
    void blit_rect(const Rect2i p_rect, const Ref<Texture2D> &p_source, const Color &p_modulate = Color(1, 1, 1, 1), int p_mipmap = 0, const Ref<Material> &p_material = Ref<Material>());
    void blit_rect_multi(const Rect2i p_rect, const TypedArray<Texture2D> &p_sources, const TypedArray<DrawableTexture2D> &p_extra_targets, const Color &p_modulate = Color(1, 1, 1, 1), int p_mipmap = 0, const Ref<Material> &p_material = Ref<Material>());

   void generate_mipmaps();
};
```

I made a detailed listing of the  modified files in a spreadsheet [here](https://docs.google.com/spreadsheets/d/1SR8Irf4NM7dqYEHa2qnSMqH9q2wZIY-_SLiEfz8grFA/edit?usp=sharing)

---

*Bugsquad edit:* Implements and fixes https://github.com/godotengine/godot-proposals/issues/7379

Commit:
DrawableTextures

Implementing DrawableTextures based on: https://github.com/godotengine/godot-proposals/issues/7379

BastiaanOlij commented on drivers/gles3/storage/texture_storage.cpp:
```diff
@@ -1370,6 +1594,15 @@ Vector<Ref<Image>> TextureStorage::texture_3d_get(RID p_texture) const {
 }
 
+void TextureStorage::texture_drawable_generate_mipmaps(RID p_texture) {
+	// TODO: Generate mipmaps code
```
Just call [glGenerateMipmap](https://registry.khronos.org/OpenGL-Refpages/gl4/html/glGenerateMipmap.xhtml) ;)

BastiaanOlij commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -1414,6 +1586,11 @@ void TextureStorage::texture_proxy_update(RID p_texture, RID p_proxy_to) {
 
+// Output textures in p_textures must ALL BE THE SAME SIZE
+void TextureStorage::texture_drawable_blit_rect(const TypedArray<RID> &p_textures, const Rect2i &p_rect, RID p_material, const Color &p_modulate, const TypedArray<RID> &p_source_textures, int p_to_mipmap) {
+	// TODO: Write this code for Renderer_RD
```
So for RD this is a little more involved than with OpenGL but the functionality already exists. If I recall correctly it's a matter of adding some extra logic when initialising the texture to create texture views for each miplayer.

Then here you can loop through the layers and use the copyeffect to scale down each layer. There might even be a function that already handles this loop for you. I'll see if I can find an example of this, pretty sure there are a few examples around the radiance map code.

BastiaanOlij commented on servers/rendering/shader_language.cpp:
```diff
@@ -390,7 +394,10 @@ const ShaderLanguage::KeyWord ShaderLanguage::keyword_list[] = {
 	{ TK_HINT_NORMAL_ROUGHNESS_TEXTURE, "hint_normal_roughness_texture", CF_UNSPECIFIED, {}, {} },
 	{ TK_HINT_DEPTH_TEXTURE, "hint_depth_texture", CF_UNSPECIFIED, {}, {} },
-
+	{ TK_HINT_BLIT_SOURCE, "hint_blit_source", CF_UNSPECIFIED, {}, {} },
```
This is just thinking out load, there might simply be too much of an issue in the shader compiler not being able to handle this. But could we just have one hint (e.g. `hint_blit_source`) and have it work on a first come, first serve basis?
E.g. the first uniform with `hint_blit_source` becomes our first source, the second uniform becomes our second source, etc.
With a compiler error if more than 4 exist.

ColinSORourke replied:
If I understand correctly, the relevant code to try that approach is in shader_compiler.cpp, 

_dump_node_code, lines 931->939

Which at a cursory glance only operates on a single token or line at a time? Those lines would need to read/write to some context variable that says how many 'sources' have already been set, and I'm unfamiliar with whether that function is supposed to access any outside context.

BastiaanOlij commented:
I had a first cursory glance at the implementation, already witnessing this being build as I was helping @ColinSORourke out, in broad strokes this looks really good. I need time to dive into the detail and do some proper stress testing with it and see if I can find any issues.

@RodZill4 , you might want to take a look at this as well, I think this would be very valuable for Material Maker. That would be a good stress test for writing to the 4 textures as I can imagine that generating texture ORMs this way could be nice.

@ColinSORourke one thing I'm wondering about as well, but this may be worth a follow up PR, instead of supplying 4 drawable textures to write too, we should also support supplying a texture array of up to 4 layers.

ColinSORourke commented:
> @ColinSORourke one thing I'm wondering about as well, but this may be worth a follow up PR, instead of supplying 4 drawable textures to write too, we should also support supplying a texture array of up to 4 layers.

Just making sure I understand correctly, do you mean
A) To change the code at the base, GLSL layer, to support `uniform sampler2D name[4]` instead of 4x `uniform sampler2D`
OR
B) to change the code at the user exposed abstraction to support passing a TextureLayered resource (AKA Texture2DArray) as a parameter instead of TypedArray<Texture2D>

BastiaanOlij commented:
> B) to change the code at the user exposed abstraction to support passing a TextureLayered resource (AKA Texture2DArray) as a parameter instead of TypedArray<Texture2D>

This one, though it's really an edge case we might want to reserve until we have a proper use case for it. I can see there might be some benefits to doing so when manipulating stereo images.

Bonkahe commented:
After some testing on the example project, firstly awesome work! second are uniforms meant to be unsupported? I tested around with sending custom data to COLOR instead of the blited texture, and one thing I tried was a uniform of source color, and that threw an error.
Shader in question:

`
shader_type texture_blit;
render_mode blend_mix;

uniform vec4 ColorMod : source_color;

uniform sampler2D source_texture : hint_blit_source;
uniform sampler2D source_texture2 : hint_blit_source2;
uniform sampler2D source_texture3 : hint_blit_source3;
uniform sampler2D source_texture4 : hint_blit_source4;

void blit() {
	// Copies from each whole source texture to a rect on each output texture.
	//COLOR = texture(source_texture, UV) * MODULATE;
	//COLOR = vec4(UV.x, UV.y, 0.0, 1.0);
	COLOR = ColorMod;
	COLOR2 = texture(source_texture2, UV);
	COLOR3 = texture(source_texture3, UV);
	COLOR4 = texture(source_texture4, UV);
}
`

Errors:
`
E 0:00:01:684   DrawTexture.gd:32 @ blitRandom(): TexBlitShaderGLES3: Fragment shader compilation failed:
0(72) : error C1503: undefined variable "m_ColorMod"

  <C++ Source>  drivers\gles3\shader_gles3.cpp:265 @ _display_error_with_code()
  <Stack Trace> DrawTexture.gd:32 @ blitRandom()
`
`
E 0:00:01:684   DrawTexture.gd:32 @ blitRandom(): Method/function failed.
  <C++ Source>  drivers\gles3\shader_gles3.cpp:407 @ ShaderGLES3::_compile_specialization()
  <Stack Trace> DrawTexture.gd:32 @ blitRandom()
`

RodZill4 commented:
> @RodZill4 , you might want to take a look at this as well, I think this would be very valuable for Material Maker. That would be a good stress test for writing to the 4 textures as I can imagine that generating texture ORMs this way could be nice.

Hmm I moved all Material Maker rendering to compute shaders, and already write to several textures at once (when painting).
I can see a use for this in other projects, though (the VR scene edit one), but will this be forward+ only?

ColinSORourke commented:
> second are uniforms meant to be unsupported? I tested around with sending custom data to COLOR instead of the blited texture, and one thing I tried was a uniform of source color, and that threw an error.

Yes they are meant to be supported, great catch. I do see the bits of GLSL I'm missing, and am going to get to work on that.

> As an aside, would it be possible to sample the current texture/destination UV we are copying into inside of the blit shader? or is that a no go? I assume it's not possible but I just wanted to be sure.

I can definitely see how that would be helpful, I don't think it would be too difficult, I'll just investigate the best way to surface those values to the user.

ColinSORourke commented:
Just pushed (4/26)
* Fix for using Uniforms from custom materials
* Fix for correct blending behaviors based on Blend Mode

Bonkahe commented:
> Just pushed (4/26)
> 
> * Fix for using Uniforms from custom materials
> * Fix for correct blending behaviors based on Blend Mode

Looks good, uniforms working just fine, as well as blend modes.

Did notice one problem, I was unable to build until I added a return value on line 1727 in `\servers\rendering\renderer_rd\storage_rd\texture_storage.cpp`

I just added this: `return RID();` obviously this is just placeholder but something to be aware of, regardless looking great!

ColinSORourke commented:
Just pushed (5/16)

Forward+ Renderer implementation! Demo Project is unchanged, you should be able to run the same demo project on either renderer by swapping Renderer in Project settings.

Note:
* Due to [this issue](https://github.com/godotengine/godot/issues/29105), Blend_Sub & Blend_Mul don't behave identically between renderer settings.

Still to-do for 'completion'
* Forward+ MipMap function
* Rigorous Testing
* Documentation 
* Feedback?

A few more talking points (I'll bring these up next Rendering Meeting as well)
* Near nothing about the actual DrawableTexture resource is handled differently than a regular Texture2D - I'm curious if 'Drawable' could just be a flag on any Texture2D instead of a distinct resource?
* Given that the Vertex Shader portion is not exposed to users, and we are not implementing blit_polygon at the moment, would it be worthwhile to add a Rotation Parameter to blit_rect, to rotate the target rect?
* A pretty common use case will be sticking with all the defaults for this - what else should the Default TextureBlit shader cover, and what should be the default setup for a DrawableTexture
* Should DrawableTextures and Blit_Rect support more Color & Blend Formats? What would need to be done for this to be allowed?

beicause commented:
I tried this PR and the demo project. It works normally.

I'm bit hard to understand the usage of `DrawableTexture2D.blit_rect_multi`. Why it is needed instead of calling `DrawableTexture2D.blit_rect` on each DrawableTexture2D?

Also, why does DrawableTexture2D only accept 4 sources instead of using them as material uniforms?

ColinSORourke commented:
> I'm bit hard to understand the usage of `DrawableTexture2D.blit_rect_multi`. Why it is needed instead of using 4 drawable textures separately and calling `DrawableTexture2D.blit_rect` on each?

Performance reasons - faster to make a single call to the GPU with 8 textures, than 4 calls to the GPU with 2 textures each.

> Also why does DrawableTexture2D only accept 4 sources instead of using them as material uniforms? Custom uniform textures seem don't work and are always black.

Extra Material Uniforms with textures should work - I'll investigate and fix.

ColinSORourke commented:
Just Pushed (5/17)

Fixes for binding custom Texture Uniforms from User Materials.

@beicause

Calinou commented:
> Erasing from textures is nearly impossible in whole of Godot sans using blend_mode substract which is sadly unusable in my case.

You may be interested in https://github.com/godotengine/godot/pull/102366, which exposes many more blend modes.

beicause commented:
I think we could expose a argument like RDPipelineColorBlendState (or a new data class for this) https://github.com/godotengine/godot-proposals/issues/12443#issuecomment-2899457162. So you can do custom blend by recreating the pipeline instead of creating a new shader.

beicause commented:
Personally I disagree with the current design of drawable textures. See also my implemention in #106331.
From my perspective, it needs to be simplified:
1. Input colors could simply be provided via uniforms.
2. Only a single output is necessary.
3. Repeatly drawing seems less essential, as it generally can be worked around.

Regardless, I truly appreciate the effort in this PR, and it was what inspired me to open #106331. However I believe its design requires careful consideration.

KoBeWi commented:
Looks like Blit shader is not properly supported by VisualShaders.
<img width="198" height="99" alt="image" src="https://github.com/user-attachments/assets/c3d8fc6a-335b-4236-a78c-fdb942fea0c2" />

KoBeWi commented on editor/icons/DrawableTexture2D.svg:
Also needs a new icon.

KoBeWi commented on scene/resources/blit_material.cpp:
```diff
@@ -0,0 +1,193 @@
+SelfList<BlitMaterial>::List BlitMaterial::dirty_materials;
+HashMap<BlitMaterial::MaterialKey, BlitMaterial::ShaderData, BlitMaterial::MaterialKey> BlitMaterial::shader_map;
+
+void BlitMaterial::init_shaders() {
```
?
Unnecessary.

KoBeWi commented on scene/resources/blit_material.h:
```diff
@@ -0,0 +1,108 @@
+		int users = 0;
+	};
+
+	static HashMap<MaterialKey, ShaderData, MaterialKey> shader_map;
```
```suggestion
	static inline HashMap<MaterialKey, ShaderData, MaterialKey> shader_map;
```
And remove definition in cpp.
Some with other static fields.

Calinou reviewed (commented):
Tested locally, it works as expected on Compatibility. This is a great start, but I think we can push further on usability, especially for less programmer-centric workflows such as VFX.

**Testing project:** [test_pr_105701.zip](https://github.com/user-attachments/files/21640832/test_pr_105701.zip)
*Press <kbd>K</kbd> to toggle real-time texture updates to compare performance.*

https://github.com/user-attachments/assets/4d721cad-a0e9-4065-81c2-24d00f8ce6c0

*Original shader: [Colorful Bubbles](https://www.shadertoy.com/view/4d3BWH) by JuliaPoo*

Some feedback:

- It currently doesn't work on Forward+/Mobile, but this is expected given what's said in OP.
- It took me a long time to figure out how you associate the shader with the DrawableTexture resource. I originally expected there would be a ShaderrMaterial property on the DrawableTexture, where you could assign a material (with its uniforms) and its shader.
  - The width/height, format and mipmaps could be set directly as properties on the DrawableTexture inspector. Combine this with a property to set an array of base textures and a ShaderMaterial, and you wouldn't need any script for basic usage. More advanced usage (e.g. blitting over an area that isn't the whole texture size) would still warrant a script, but the use case for creating procedural textures would be simplified a lot.

  - I think we can provide this high-level functionality and keep low-level behavior intact, but we need to make sure it's not initializing textures twice in this case (e.g. by reading inspector properties then the user's own `start()` call).

  - Additionally, there is no quick way to just use a plain background color to blit over the entire area to apply your own shader. Some procedural textures initialize/update asynchronously, so using GradientTexture2D won't work unless you `await get_tree().process_frame`, which is not possible if you attach the script directly to the DrawableTexture2D like I originally did (since it's a Resource and can't access SceneTree directly). I ended up attaching the script to a node instead so I could use `_process()`, but I figured it was worth mentioning nonetheless. The testing project also includes a pure black `black.png` image for this purpose.

- Having a DrawableTexture already set on a material property won't update it if you call `start()` to initialize it. You need to set it in a script again after calling `start()` for it to be recognized:

```gdscript
	drawable_texture.setup(64, 64, DrawableTexture2D.DRAWABLE_FORMAT_RGBA8, true)
	drawable_texture.blit_rect(Rect2(0, 0, 64, 64), preload("res://white.png"), Color.WHITE, 0, preload("res://drawable_texture_material.tres"))
	material_override.albedo_texture = drawable_texture  # This line is required for the texture to appear, even if we loaded it in the inspector previously.
```

If you update the texture in real-time, you also need the last line to be in `_process()`, which is more resource-intensive than it could be.

- Passing `null` as a texture parameter in `blit_rect()` causes a crash:

```
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.5.beta.custom_build (faf3a2d29b8775eb2354e339d353b03869041aa0)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
[1] /lib64/libc.so.6(+0x19c30) [0x7f3859827c30] (??:0)
[2] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0x8d917b4] (/home/hugo/Documents/Git/godotengine/godot/scene/resources/drawable_texture_2d.cpp:98 (discriminator 1))
[3] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0x8f7fda6] (/home/hugo/Documents/Git/godotengine/godot/core/variant/binder_common.h:223 (discriminator 9))
[4] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0x8f58f99] (/home/hugo/Documents/Git/godotengine/godot/core/variant/binder_common.h:371)
[5] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0x8f38610] (/home/hugo/Documents/Git/godotengine/godot/core/object/method_bind.h:337 (discriminator 1))
[6] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0x47e7288] (/home/hugo/Documents/Git/godotengine/godot/modules/gdscript/gdscript_vm.cpp:2054 (discriminator 1))
[7] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0x4574442] (/home/hugo/Documents/Git/godotengine/godot/modules/gdscript/gdscript.cpp:192)
[8] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0x4575e88] (/home/hugo/Documents/Git/godotengine/godot/modules/gdscript/gdscript.cpp:425)
[9] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0xb2ac980] (/home/hugo/Documents/Git/godotengine/godot/core/object/object.cpp:1019 (discriminator 2))
[10] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0xb2a8730] (/home/hugo/Documents/Git/godotengine/godot/core/object/object.cpp:314)
[11] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0x90c3ec4] (/home/hugo/Documents/Git/godotengine/godot/core/string/string_name.h:182)
[12] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0x90c837b] (/home/hugo/Documents/Git/godotengine/godot/scene/resources/resource_format_text.cpp:1412 (discriminator 1))
[13] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0xab91a4f] (/home/hugo/Documents/Git/godotengine/godot/core/io/resource_loader.cpp:314 (discriminator 1))
[14] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0xab92604] (/home/hugo/Documents/Git/godotengine/godot/core/io/resource_loader.cpp:396 (discriminator 7))
[15] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0xab93675] (/home/hugo/Documents/Git/godotengine/godot/core/io/resource_loader.cpp:663)
[16] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0x90c14f6] (/home/hugo/Documents/Git/godotengine/godot/scene/resources/resource_format_text.cpp:463 (discriminator 5))
[17] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0x90c837b] (/home/hugo/Documents/Git/godotengine/godot/scene/resources/resource_format_text.cpp:1412 (discriminator 1))
[18] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0xab91a4f] (/home/hugo/Documents/Git/godotengine/godot/core/io/resource_loader.cpp:314 (discriminator 1))
[19] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0xab92604] (/home/hugo/Documents/Git/godotengine/godot/core/io/resource_loader.cpp:396 (discriminator 7))
[20] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0xab93675] (/home/hugo/Documents/Git/godotengine/godot/core/io/resource_loader.cpp:663)
[21] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0xab92f95] (/home/hugo/Documents/Git/godotengine/godot/core/io/resource_loader.cpp:555)
[22] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0x3caab3a] (/home/hugo/Documents/Git/godotengine/godot/main/main.cpp:4555 (discriminator 1))
[23] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0x3ba3299] (/home/hugo/Documents/Git/godotengine/godot/platform/linuxbsd/godot_linuxbsd.cpp:84 (discriminator 1))
[24] /lib64/libc.so.6(+0x35f5) [0x7f38598115f5] (??:0)
[25] /lib64/libc.so.6(__libc_start_main+0x88) [0x7f38598116a8] (??:0)
[26] /home/hugo/Documents/Git/godotengine/godot/bin/godot.linuxbsd.editor.x86_64() [0x3ba30a5] (??:?)
-- END OF C++ BACKTRACE --
```

- `COLOR` builtins are named `COLOR`/`COLOR2`/`COLOR3`/`COLOR4`, but sky shaders go for different numbering as per the [documentation](https://docs.godotengine.org/en/stable/tutorials/shaders/shader_reference/sky_shader.html#global-built-ins):

> There are 4 `LIGHTX` lights, accessed as `LIGHT0`, `LIGHT1`, `LIGHT2`, and `LIGHT3`.

I'd go for `COLOR0`/`COLOR1`/`COLOR2`/`COLOR3` for consistency.

- DrawableTexture should get its own icon, so it can look different from Texture2D (which is an abstract class, and therefore has a gray icon):

<img width="279" height="573" alt="Screenshot_20250806_225450" src="https://github.com/user-attachments/assets/8564759a-6b6f-484f-a99d-3169f2cb1b61" />

For future PRs:

- `TIME` is not accessible in `texture_blit` shaders, but I think it should to make procedural effects easier. You can emulate it for now by passing a `time` uniform manually in `_process()`, then calling `setup()` every frame.

  - Having this as a built-in implementation would also need an update mode property in DrawableTexture, which would control when to update the texture: **Automatic** (always update only if `TIME` is used), **Once**, **Always**.

TokisanGames commented:
How do we specify that textures in a texture array are to be drawable?

beicause commented:
> How do we specify that textures in a texture array are to be drawable?

I think it's the same as texture rd, but with some extra usage bits (color_attachment, can_copy_to or storage) for drawing.

clayjohn commented:
> How do we specify that textures in a texture array are to be drawable?

Drawable textures are currently only Texture2Ds. 

A texture array is a different texture type entirely

arkology commented:
It doesn't seem to be working at the moment, does it?
|  <img width="1152" height="1000" alt="godot windows editor x86_64_anmW0h8lb3" src="https://github.com/user-attachments/assets/776b219f-d577-49c4-bc08-1ba98aa21de3" /> |  <img width="1152" height="1000" alt="godot windows editor x86_64_iZ6tpIgpk5" src="https://github.com/user-attachments/assets/b05de6db-89b6-44b2-bc7b-2fd74d08d80b" /> |
| --- | --- |

ColinSORourke commented:
@arkology Finally got a chance to get to this (moved out of a house this week) - and just pushed a build that should fix this.

Demo project was fine, I couldn't recreate the error, but that must be my machine not being fussy about the implicit int to float conversion. I removed that implicit conversion from the Shader, so the Demo Project should work for you now, let me know if it still doesn't. Also synced the PR to Master!

arkology commented:
Thanks for updates!

Please note that gdshader file in test project throws errors (although I see from the discussion why they appeared).

> Draws to the given rectangle on this texture, as well as on up to 3 extra target DrawableTexture. All extra targets must be the same size and DrawableFormat as the original target, otherwise the Shader may fail. Expects up to 4 Source Textures, but will replace missing Sources with default Black Textures.

Maybe there should be a warning if array size is bigger?

I had a idea that `DrawableTexture` could be used as `AtlasTexture` wrapper for `PointLight2D` (`PointLight2D` ignores region of `AtlasTexture`), but failed to make it work. 
Is it possible to create fully transparent `DrawableTextures`? Right now it is always white rect.
And "DrawableTexture" completely ignores the "AtlasTexture" region, and this is a big problem for me.

beicause commented:
>Is it possible to create fully transparent DrawableTextures? Right now it is always white rect.

It's possible, but you need to draw a fully transparent rect first manually.

>And "DrawableTexture" completely ignores the "AtlasTexture" region, and this is a big problem for me.

As far as I know, AtlasTexture has limited use case, it should works in common CanvasItem drawing,  but it doesn't work in some places and 3D.

By the way, I hope that when testing this PR, please also compare it with #106331 (as they follow different designs) to gather more feedback on this. If everyone reaches a consensus, I can close my PR.

arkology commented:
> but it doesn't work in some places and 3D.

Yes, I know that, `AtlasTexture` support should be added into engine source code manually. And I hope that "AtlasTexture" support can be added to "DrawableTexture".

ColinSORourke commented:
> Maybe there should be a warning if array size is bigger?

You're suggesting a warning like "Passed 5 inputs into Blit_Rect_Multi, only first 4 will be used"? I could see that, but not sure how useful it would be and if it's common in Godot to throw warnings for excess input.

> And "DrawableTexture" completely ignores the "AtlasTexture" region, and this is a big problem for me.

To be clear, you're suggesting a workflow where you set the 'atlas' of an AtlasTexture as a DrawableTexture, like 
```
var myAtlas: AtlasTexture = AtlasTexture.new()
myAtlas.atlas = drawTex
```
And the issue is that if you then call `myAtlas.atlas.blit_rect(rect, source)` , the rect passed in to blit_rect operates from the top left corner of the whole texture, and not the top_left corner of the atlas's current 'region'? I'm unsure how we could codify that behavior in-engine, but you can run some math on rect & region to achieve the desired effect I think. Something like:
```
myAtlas.atlas.blit_rect(Rect2i(myAtlas.region.position + rect.position, rect.size), source)
```

> Is it possible to create fully transparent DrawableTextures? Right now it is always white rect.
> It's possible, but you need to draw a fully transparent rect first manually.
> This sounds not very convenient :D
> I am at "small brain mode" at the moment, but looking at DrawableTexture2D docs description as a user, blend_mode will be really useful and frequently-used IMO.

Yeah, to get a transparent DrawableTexture, you have to blit the whole rect with a transparent Source on Blend_Mode_Disabled. Currently the most straightforward code for that would be
```
var myMat = BlitMaterial.new()
myMat.blend_mode = BlitMaterial.BLEND_MODE_DISABLED
drawTex.blit_rect(whole_rect, transparent_texture, Color.WHITE, 0, myMat)
```

We've had conversations about simpler ways to expose other blend modes as easier to write default behaviors, but for various reasons are waiting for more feedback - I appreciate yours.

ColinSORourke commented:
> About `AtlasTexture` - try to create `AtlasTexture` resource with some region in your test project and place the path to it into one of load() in test project script. It will draw the whole sprite. 

Yeah I see the issue and just programmed the obvious fix for it, but the obvious fix might be costly for memory. The issue lies more with the under-the-hood of AtlasTexture than Drawable, I'll bring it up in the next meeting to talk about if there's a better solution, and if we want to move forward with a solution at all

beicause commented:
AtlasTexture is merely a wrapper for `CanvasItem.draw_texture_region`. Note that the underlying data is still the entire texture.

> But looking at the DrawableTexture2D documentation description as a user, blend_mode will be really useful and frequently-used in my opinion.

I agree that blend mode is commonly used and think it should be implemented as a method parameter rather than a shader mode. Additionally, drawing a full-rect color can be achieved through `clear_color`.

BastiaanOlij reviewed (approved):
@ColinSORourke @clayjohn OK, code wise I think this all looks good, nothing stands out to me that needs touching up. The only thing really is that the `linear_to_srgb` conversion is an all or nothing deal. In the sample project that I made it's important that only SOURCE0 is converted. However by setting the correct format I think it worked correctly anyway.

So I created a sample project, a very bare bones texture painter that paints albedo/normal/ao/roughness and metallic. It all worked very well and did exactly what it should do.
You can find the resulting demo project here: https://github.com/BastiaanOlij/drawable-textures-demo

I also streamed developing this, it's a 2 hour watch but might help provide some more context: https://www.youtube.com/live/O1gwSoXTv_g

There is room for improvement but the things I ran into can all be follow up things, the heart of this is solid and I'm for merging the implementation as is.
The things I think we should consider for the future are:
- I missed the ability to set a default color for the content. White is not a good default for a normal map or ORM texture :)
- I know Reduz was adamant around the `setup` function and doing everything in code, but working with it I feel it makes far more sense to make these properties of the texture.
- The aforementioned issues with wanting more control/certainty about linear output v.s. sRGB output.

ColinSORourke commented:
Just push a FINAL set of changes. Patch Notes:
* Changed `DrawableTexture2D.setup` API so the user can specify a base color to initialize the texture to. This will unfortunately error out some of the cool projects built earlier in this thread, but can easily be fixed adjusting the GDScript.
* Improved SRGB handling - SRGB now appropriately adjusts by output in Blit_Rect_Multi, instead of being all or nothing.
* Changed the 'fallback' texture for blit_sources to white instead of black. This makes it easier to do `Texture.blit_rect(rect, null, modulateColor)` as a simple way to blit a flat color.
* Added Experimental warning to DrawableTexture2D Docs - still highly sensitive to API feedback on those functions.
* Adjusted the Blend_Sub and Blend_Add behaviors - should better match intuition.
* Simplified BlitMaterial to recompile Shader immediately (like SkyMaterials) rather than going through a queuing process
* Rebased o Master (Sep. 25th)

@clayjohn @BastiaanOlij Shouldn't need a deep second review - The SRGB handling and BlitMaterial were the most complex changes, and I did my best to verify I didn't break anything. Should be able to move forward for 4.6 Dev 2.

I also built a new Demo Project - good for testing behaviors and showcasing the functionalities of this PR :)
[drawable-texture-tests.zip](https://github.com/user-attachments/files/22552003/drawable-texture-tests.zip)
It's got 5 scenes:
* A basic painting scene using simple DrawableTexture
* A Scene showcasing Blit_Rect_Multi behavior, blitting to 4 targets of differing formats
* A Scene showcasing MipMap behavior, where a user can draw directly to a lower mipmap level
* A Scene showcasing the 5 Blend Mode behaviors
* A Scene showcasing a fancy Texture_Blit shader, modified from @Calinou's example

BastiaanOlij commented:
> @clayjohn @BastiaanOlij Shouldn't need a deep second review - The SRGB handling and BlitMaterial were the most complex changes, and I did my best to verify I didn't break anything. Should be able to move forward for 4.6 Dev 2.

Agreed from my end, especially if my demo project is working fine. I think we've clearly discussed all of it.
@clayjohn  might want to do a final pass over but IMHO, this would be a great merge for dev2.

beicause commented:
I got Vulkan validation errors when running `MipMapDemo.tscn`(the same as https://github.com/godotengine/godot/pull/110076#issuecomment-3322387825):
```
Vulkan 1.4.318 - Forward+ - Using Device #0: AMD - AMD Radeon Graphics (RADV RENOIR)

ERROR: Vulkan Debug Report: object - 2048390162548551
vkCreateFramebuffer(): pCreateInfo->pAttachments[0] has mip levelCount of 9 but only a single mip level (levelCount ==  1) is allowed when creating a Framebuffer.
The Vulkan spec states: If flags does not include VK_FRAMEBUFFER_CREATE_IMAGELESS_BIT, each element of pAttachments must only specify a single mip level (https://docs.vulkan.org/spec/latest/chapters/renderpass.html#VUID-VkFramebufferCreateInfo-pAttachments-00883)
   at: _debug_report_callback (drivers/vulkan/rendering_context_driver_vulkan.cpp:665)
WARNING: Vulkan Debug Report: object - 93944788457168
vkCmdDispatch(): the storage image descriptor [VkDescriptorSet 0x6e300000006e3, Set 3, Binding 0, Index 0] is accessed by a OpTypeImage that has a Format operand Rgba16f (equivalent to VK_FORMAT_R16G16B16A16_SFLOAT) which doesn't match the VkImageView 0x6e100000006e1 format (VK_FORMAT_R8G8B8A8_UNORM). Any loads or stores with the variable will produce undefined values to the whole image (not just the texel being accessed).
Spec information at https://docs.vulkan.org/spec/latest/chapters/textures.html#textures-format-validation
     at: _debug_report_callback (drivers/vulkan/rendering_context_driver_vulkan.cpp:662)
```
Also the texture doesn't need `TEXTURE_USAGE_STORAGE_BIT` in Mobile renderer.

ColinSORourke commented:
@BastiaanOlij @beicause Thanks for the catch on this Vulkan error. I don't have a Vulkan device I can test on, but I just pushed a fix which I believe should resolve it based on Bastiaan's comment.

<img width="1269" height="319" alt="Screenshot 2025-09-27 at 9 12 50 AM" src="https://github.com/user-attachments/assets/8e3f52d1-57e3-47a2-84de-37b8714b1dfa" />

Tar_textures get bound directly later in Blit_Rect, but previously I only created the view if p_to_mipmap was not 0, and now we always call create_shared from slice. This is in the Forward+ render, as far as I can tell my Compatibility version always binds to a specific mipmap view. Let me know if this resolves the issue, it at least didn't break MipMapDemo.tscn on my machine.

ColinSORourke commented:
@beicause Just pushed - Also addressed Mobile Render Mipmaps, another good catch. I'll admit my Mobile Renderer testing hasn't been as thorough. 

<img width="859" height="285" alt="Screenshot 2025-09-27 at 2 04 37 PM" src="https://github.com/user-attachments/assets/ddaeec53-f228-41a2-85d9-5e2c81f2d3ed" />

akien-mga commented:
There are some style issues to fix (missing `\m` at EOF): https://github.com/godotengine/godot/actions/runs/19780756432/job/56680572703?pr=105701

ColinSORourke commented:
Great catch @mk56-spn - I investigated today and determined that this is because Blit_Rect overflows RenderingDevice's texture_owner, because every blit_rect call creates new shared textures as the targets for the frame buffers.

No solution yet, but obviously these temporary shared textures need to be freed at some point, rather than building up infinitely. The obvious solution caused my Frame Buffer creation to crash because I freed the Shared before the draw call actually processed - going to keep investigating, will have a fix soon!

mk56-spn commented:
> Great catch @mk56-spn - I investigated today and determined that this is because Blit_Rect overflows RenderingDevice's texture_owner, because every blit_rect call creates new shared textures as the targets for the frame buffers.
> 
> No solution yet, but obviously these temporary shared textures need to be freed at some point, rather than building up infinitely. The obvious solution caused my Frame Buffer creation to crash because I freed the Shared before the draw call actually processed - going to keep investigating, will have a fix soon!

Interesting, overall very solid implementation i must say!

say is there any chance we could somehow, blit at scene render time to access scree textures and the like, or is the pipeline not feasible for that

ColinSORourke commented:
Resolved @mk56-spn reported issue in a push just now (12/22). Ran his test case with the fix for 40 minutes at 120fps (so approximately 864k Blit_rect_multi calls), no issues encountered.

Fix, per @clayjohn's suggestion on Rocket Chat, is to pre-generate the slices and store them as part of the DrawableTexture. These are stored in a new Vector<RID> in the TextureStorage::Texture struct, cached_rd_slices, and are initialized as part of Drawable_Texture_initialize. They do not need a separate free process, as they are freed as dependencies of RD_Texture when it is freed.

clayjohn commented on doc/classes/DrawableTexture2D.xml:
```diff
@@ -0,0 +1,71 @@
+	<tutorials>
+	</tutorials>
+	<methods>
+		<method name="blit_rect" experimental="This function and its parameters are likely to change in the 4.6 Dev Cycle">
```
```suggestion
		<method name="blit_rect" experimental="This function and its parameters are likely to change in the 4.7 Dev Cycle">
```

clayjohn commented on doc/classes/DrawableTexture2D.xml:
```diff
@@ -0,0 +1,71 @@
+				Draws to given [param rect] on this texture by copying from the given [param source]. A [param modulate] color can be passed in for the shader to use, but defaults to White. The [param mipmap] value can specify a draw to a lower mipmap level. The [param material] parameter can take a ShaderMaterial with a TextureBlit Shader for custom drawing behavior.
+			</description>
+		</method>
+		<method name="blit_rect_multi" experimental="This function and its parameters are likely to change in the 4.6 Dev Cycle">
```
```suggestion
		<method name="blit_rect_multi" experimental="This function and its parameters are likely to change in the 4.7 Dev Cycle">
```

clayjohn reviewed (changes requested):
Just some nitpicks and a rebase needed, then let's merge for 4.7 dev1!

clayjohn reviewed (approved):
Looks great! Thank you so much for your persistence, this has been a huge amount of work, but the results are looking great. Let's get this merged in dev1 so we can get lots of feedback prior to stable!!!

Repiteo commented:
Thanks!

gnat commented:
Using this in an upcoming game right now! Thank you for the wonderful work @ColinSORourke
